### PR TITLE
Topic/parameter server reduced model

### PR DESCRIPTION
**Pull request by olivier-stasse**
This PR fixes #131 

With Commit c239120 it provides objects which can handle trees of parameters. The trees are indexed by names. For now only basic types can be registered:
 * int
 * bool
 * double
 * string

The two objects are : ParameterServer which is an Entity providing python bindings and access to the map of parameter trees. A parameter tree is handled through a pure C++ object. 

The default parameter tree is named 'robot'. It is populated in dynamic_graph_bridge when used with ROS ( more precisely here https://github.com/stack-of-tasks/dynamic_graph_bridge/blob/6bce07ce8ea31bf5226aa08f41c896851e0e9a1b/src/sot_loader_basic.cpp#L46 ). More precisely it copy the **/robot_description** ros parameter and set it in the tree named 'robot' under the parameter name **/robot_description'**.

The RobotUtil tree can be searched for through the **sot::getRobotUtil()** function. It can be used to build a reduced model with pinocchio:
https://github.com/olivier-stasse/sot-pattern-generator/blob/5a1358b3a56a27903ef4eeeea1e2e4eba6aa110f/src/pg.cpp#L614

For now the inconvenient of this PR, is the fact that on 18.04 LTS the get_parameter method of RobotUtils can not handle const Type & calls. (Generating a seg fault)

### CODE REVIEW

**Comment by olivier-stasse** [comment]
I will update the PR with the rewrote history.

File: include/sot/core/parameter-server.hh [hunk 1]
```diff
@@ -99,9 +100,36 @@ public:
...
-                    const std::string &ParameterValue);
-  std::string getParameter(const std::string &ParameterName);
+                    const Type &ParameterValue)
+  {
+    if (!m_initSucceeded) {
+      std::ostringstream oss;
+      oss << ParameterValue;
+      SEND_WARNING_STREAM_MSG("Cannot set parameter " + ParameterName + " to " +
+                              oss.str().c_str() +
+                              " before initialization!");
```
Comment (jmirabel) [inline include/sot/core/parameter-server.hh:112]: What about the following one-liner ?
```suggestion
      DYNAMIC_GRAPH_ENTITY_WARNING(*this) << "Cannot set parameter " << ParameterName
        << " to " << ParameterValue << " before initialization!";
```

File: include/sot/core/parameter-server.hh [hunk 2]
```diff
@@ -35,6 +35,7 @@
 /* --- INCLUDE --------------------------------------------------------- */
 /* --------------------------------------------------------------------- */
 
+#include <sstream>
```
Comment (jmirabel) [inline include/sot/core/parameter-server.hh:38]: If you apply the one-liner below:
```suggestion
```

File: include/sot/core/parameter-server.hh (hunk 1 above)
Comment (jmirabel) [reply include/sot/core/parameter-server.hh:112]: Note that I removed the stream. I don't think it makes sense here.

File: include/sot/core/robot-utils.hh [hunk 3]
```diff
@@ -11,6 +11,28 @@
...
+
+#include <boost/property_tree/ptree.hpp>
+
+#ifdef BOOST_MPL_LIMIT_VECTOR_SIZE_PUSH
+#pragma pop_macro("BOOST_MPL_LIMIT_VECTOR_SIZE")
+#endif
+
+#ifdef BOOST_MPL_LIMIT_LIST_SIZE_PUSH
+#pragma pop_macro("BOOST_MPL_LIMIT_LIST_SIZE")
+#endif
```
Comment (jmirabel) [inline include/sot/core/robot-utils.hh:34]: Can you put the rationale for adding all this preprocessor commands ?

File: include/sot/core/robot-utils.hh [hunk 4]
```diff
@@ -217,28 +259,74 @@ public:
...
+    }
+    catch(const boost::property_tree::ptree_error &e)
+    {
+      std::ostringstream oss;
+      oss << "Robot utils: parameter path is invalid " << '\n'
+          << " for set_parameter("
+          << parameter_name << ")\n"
+          << e.what() << std::endl;
+      sendMsg(oss.str(),
+              MSG_TYPE_ERROR);
```
Comment (jmirabel) [inline include/sot/core/robot-utils.hh:280]:
```suggestion
      DYNAMIC_GRAPH_ENTITY_ERROR(*this) << "Robot utils: parameter path is invalid " << '\n'
          << " for set_parameter("
          << parameter_name << ")\n"
          << e.what();
```

File: src/tools/robot-utils.cpp [hunk 5]
```diff
@@ -5,28 +5,92 @@
...
+  return mimic_joints_;
 }
 
+void ExtractJointMimics::go_through_full()  {
+  current_joint_name_="";
+  go_through(tree_,0,0);
+}
+
+void ExtractJointMimics::go_through(pt::ptree &pt, int level, int stage)
+{
```
Comment (jmirabel) [inline src/tools/robot-utils.cpp:58]: Might be worth adding a comment on what happens in this recursive function.

File: include/sot/core/robot-utils.hh [hunk 6]
```diff
@@ -217,28 +259,74 @@ public:
...
+    }
+    catch(const boost::property_tree::ptree_error &e)
+    {
+      std::ostringstream oss;
+      oss << "Robot utils: parameter path is invalid " << '\n'
+          << " for get_parameter("
+          << parameter_name << ")\n"
+          << e.what() << std::endl;
+      sendMsg(oss.str(),
+              MSG_TYPE_ERROR);
```
Comment (jmirabel) [inline include/sot/core/robot-utils.hh:310]:
```suggestion
      DYNAMIC_GRAPH_ENTITY_ERROR(*this) << "Robot utils: parameter path is invalid " << '\n'
          << " for get_parameter("
          << parameter_name << ")\n"
          << e.what() << '\n';
```

File: include/sot/core/robot-utils.hh (hunk 4 above)
Comment (olivier-stasse) [reply include/sot/core/robot-utils.hh]: I was too fast. RobotUtils class is not an entity.

File: include/sot/core/parameter-server.hh (hunk 1 above)
Comment (olivier-stasse) [reply include/sot/core/parameter-server.hh]: I cannot use the DYNAMIC_GRAPH_ENTITY_WARNING because the class RobotUtils is not an entity.

Comment (jmirabel) [reply include/sot/core/parameter-server.hh]: The class here is ParameterServer, which is an entity.

File: include/sot/core/robot-utils.hh (hunk 4 above)
Comment (jmirabel) [reply include/sot/core/robot-utils.hh]: Here, it is not an entity, so you can use `dgRTLOG() << ...`

File: include/sot/core/robot-utils.hh (hunk 3 above)
Comment (olivier-stasse) [reply include/sot/core/robot-utils.hh]: Done in commit f7f8ebb

File: src/tools/robot-utils.cpp (hunk 5 above)
Comment (olivier-stasse) [reply src/tools/robot-utils.cpp:68]: Done in commit f7f8ebb

File: include/sot/core/parameter-server.hh (hunk 1 above)
Comment (olivier-stasse) [reply include/sot/core/parameter-server.hh]: You are right. Applied in commit f7f8ebb

**Comment by olivier-stasse** [comment]
This PR is now ready for merging. **const Type &** for **get_parameter** are changed to **Type get_parameter** as it is faulty on Ubuntu 18.04.

**Review by nim65s — CHANGES_REQUESTED** [review]
format is not happy: https://gitlab.laas.fr/stack-of-tasks/sot-core/-/jobs/89067 while it is almost OK on devel. could you run clang-format ?

That's all for me, I guess this can be included in #157 

@jmirabel : do you see anything else ?

**Comment by jmirabel** [comment]
> **const Type &** for **get_parameter** are changed to **Type get_parameter** as it is faulty on Ubuntu 18.04.

The [prototype](https://www.boost.org/doc/libs/1_74_0/doc/html/boost/property_tree/basic_ptree.html) of `ptree::get` for boost 1.74 is
```cpp
  template<typename Type> Type get(const path_type &) const;
```
so I suspect `const Type&` worked on other platforms only by luck.

**Review by jmirabel — APPROVED** [review]

**Comment by olivier-stasse** [comment]
> > **const Type &** for **get_parameter** are changed to **Type get_parameter** as it is faulty on Ubuntu 18.04.
> 
> The [prototype](https://www.boost.org/doc/libs/1_74_0/doc/html/boost/property_tree/basic_ptree.html) of `ptree::get` for boost 1.74 is
> 
> ```c++
>   template<typename Type> Type get(const path_type &) const;
> ```
> 
> so I suspect `const Type&` worked on other platforms only by luck.

Thanks for pointing this out. I was mislead by other accessors providing const Type & and did not check that get could provide it...

File: include/sot/core/robot-utils.hh [hunk 7]
```diff
@@ -22,11 +22,25 @@
...
+#endif
+
+
 #include <boost/property_tree/ptree.hpp>
+
+#ifdef UNDEF_BOOST_MPL_LIMIT_VECTOR_SIZE
 #pragma pop_macro("BOOST_MPL_LIMIT_VECTOR_SIZE")
-#else
-#include <boost/property_tree/ptree.hpp>
+#endif
```
Comment (jmirabel) [inline include/sot/core/robot-utils.hh:39]:
```suggestion
#undef UNDEF_BOOST_MPL_LIMIT_VECTOR_SIZE
#endif
```

File: include/sot/core/robot-utils.hh [hunk 8]
```diff
@@ -22,11 +22,25 @@
...
+
+#ifdef UNDEF_BOOST_MPL_LIMIT_LIST_SIZE
+#pragma pop_macro("BOOST_MPL_LIMIT_LIST_SIZE")
```
Comment (jmirabel) [inline include/sot/core/robot-utils.hh:42]:
```suggestion
#pragma pop_macro("BOOST_MPL_LIMIT_LIST_SIZE")
#undef UNDEF_BOOST_MPL_LIMIT_LIST_SIZE
```

**Review by nim65s — APPROVED** [review]

**Pull request merged**